src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+
+  CubeType output(300, 1000, 13);
+  for(size_t k = 0; k < features.n_slices; ++k)
+  {
+    size_t r_idx = 0;
+    for(size_t i = 0; i < grid_pos.n_elem; ++i) //loop length : 25
+    {
+      for(size_t j = i + 1; j < grid_pos.n_elem; ++j) //loop length : 25
+      {
+        output.slice(k).row(r_idx) = features.slice(k).row(grid_pos(i))
+                                    - features.slice(k).row(grid_pos(j));
+        ++r_idx;
+      }
+    }
+  }
+  return output;
+}
+
+//returns 300,1000,13 dimension features.
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+GetSSFtr(CubeType const &channels, arma::umat const &loc)
+{
+  const size_t shrink = this->options["shrink"];
+  const size_t p_size = this->options["p_size"] / shrink;
+
+  //n_cell: number of self similarity cells
+  const size_t n_cell = this->options["n_cell"];
+  const size_t half_cell_size = (size_t) round(p_size / (2.0 * n_cell));
+
+  arma::uvec g_pos(n_cell);
+  for(size_t i = 0; i < n_cell; ++i)
+  {
+    g_pos(i) = (size_t)round( (i + 1) * (p_size + 2 * half_cell_size \
+                     - 1) / (n_cell + 1.0) - half_cell_size);
+  }
+  arma::uvec grid_pos(n_cell * n_cell);
+  size_t k = 0;
+  for(size_t i = 0; i < n_cell; ++i)
+  {
+    for(size_t j = 0; j < n_cell; ++j)
+    {
+      grid_pos(k) = g_pos(i) * p_size + g_pos(j);
+      ++k;
+    }
+  }
+
+  CubeType wind = this->ViewAsWindows(channels, loc);
+  CubeType re_wind = this->Rearrange(wind);
+
+  return this->PDist(re_wind, grid_pos); 
+}
+
+template<typename MatType, typename CubeType>
+arma::field<CubeType> StructuredForests<MatType, CubeType>::
+GetFeatures(MatType const &image, arma::umat &loc)
+{
+  const size_t row_size = this->options["row_size"];
+  const size_t col_size = this->options["col_size"];
+  const size_t bottom = (4 - (image.n_rows / 3) % 4) % 4;
+  const size_t right = (4 - image.n_cols % 4) % 4;
+  //cout << "Botttom = " << bottom << " right = " << right << endl;
+
+  CubeType InImage(image.n_rows / 3, image.n_cols, 3);
+
+  for(size_t i = 0; i < 3; ++i)
+  {
+    InImage.slice(i) = image.submat(i * row_size, 0, \
+                      (i + 1) * row_size - 1, col_size - 1);
+  }
+  
+  CubeType OutImage = this->CopyMakeBorder(InImage, 0, 0, bottom, right);
+
+  const size_t num_channels = 13;
+  const size_t shrink = this->options["shrink"];
+  const size_t rsize = OutImage.n_rows / shrink;
+  const size_t csize = OutImage.n_cols / shrink;
+
+  /* this part gives double free or corruption out error
+     when executed for a second time */
+  CubeType reg_ch = CubeType(rsize, csize, num_channels);
+  CubeType ss_ch = CubeType(rsize, csize, num_channels);
+  this->GetShrunkChannels(InImage, reg_ch, ss_ch);
+  
+  loc /= shrink;
+
+  CubeType reg_ftr = this->GetRegFtr(reg_ch, loc);
+  CubeType ss_ftr = this->GetSSFtr(ss_ch, loc);
+  arma::field<CubeType> F(2,1);
+  F(0,0) = reg_ftr;
+  F(1,0) = ss_ftr;
+  return F;
```

Comment:
This wont be required now, I'll pass the reference of reg_ftr and ss_ftr in this function and return void. Then we could combine both type of features into single matrix, in PrepareData()